File: src/cube.js
export function createCube(scene, x, y, z) {
  // add cube to scene
  const geometry = new THREE.BoxGeometry();
  const color = Math.random() * 0xffffff;
  const material = new THREE.MeshBasicMaterial({ color });
  const cube = new THREE.Mesh(geometry, material);
  cube.position.x = x;
  cube.position.y = y;
  scene.add(cube);

  // animate
  const animate = () => {
    requestAnimationFrame(animate);
    cube.rotation.x += 0.01;
    cube.rotation.y += 0.01;
  };
  animate();
}

File: leptos_three/src/cube.rs
use crate::providers;
use crate::three::*;
use leptos::*;
use std::rc::Rc;

#[component]
pub fn Cube(
    cx: Scope,
    children: Children,
    #[prop(default = [0.0,0.0,0.0])] position: [f64; 3],
) -> impl IntoView {
    let scene = use_context::<providers::SceneContext>(cx).unwrap().0;

    let (object3d, set_object3d) = create_signal::<Option<Rc<Mesh>>>(cx, None);

    provide_context(cx, providers::Object3DContext(object3d));

    create_effect(cx, move |_| {
        if let Some(scene) = scene.get() {
            let mat = MeshBasicMaterial::new();
            let geo = BoxGeometry::new();
            let mesh = Mesh::new(&geo, &mat);
            mesh.position().set_x(position[0]);
            mesh.position().set_y(position[1]);
            mesh.position().set_z(position[2]);
            scene.add(&mesh);
            set_object3d.set(Some(Rc::new(mesh)));
        }
    });

    view! { cx,
            <>
            {children(cx)}
            </>
    }
}
